src/cose/cose.rs

```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
     Ok(())
 }
 
-/// Sign the payload and return a serialised COSE_Sign object.
-#[allow(unused_variables)]
-pub fn sign(payload: &[u8]) -> Result<Vec<u8>, CoseError> {
-    unimplemented!()
+// XXX: This works only with P256 for now!
+fn build_pkcs8(public_key: &[u8], secret_key: &[u8]) -> Vec<u8> {
+    let mut pkcs8: Vec<u8> =
+        vec![
+      0x30, 0x81, 0x87, 0x02, 0x01, 0x00, 0x30, 0x13, 0x06, 0x07, 0x2a, 0x86,
+      0x48, 0xce, 0x3d, 0x02, 0x01, 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d,
+      0x03, 0x01, 0x07, 0x04, 0x6d, 0x30, 0x6b, 0x02, 0x01, 0x01, 0x04, 0x20,
+      // Private key: len, bytes.
+      // Public key (uncompressed EC point for now): 0xa1, len, type, len, 0x00, bytes.
+      ];
+    pkcs8.extend_from_slice(secret_key);
+    pkcs8.push(0xa1);
+    let len = (public_key.len() + 4 - 1) as u8;
+    pkcs8.push(len);
+    pkcs8.push(0x03);
+    pkcs8.push(len - 2);
+    pkcs8.push(0x00);
+    pkcs8.extend_from_slice(public_key);
+
+    return pkcs8;
+}
+
+/// RFC 8152 Section 4.4
+///    Sig_structure = [
+///        context : "Signature" / "Signature1" / "CounterSignature",
+///        body_protected : empty_or_serialized_map,
+///        ? sign_protected : empty_or_serialized_map,
+///        external_aad : bstr,
+///        payload : bstr
+///    ]
+fn build_signature_structure(payload: &[u8], alg: CoseSignatureType) -> Vec<u8> {
+    let mut sig_structure: Vec<CborType> = Vec::new();
+    sig_structure.push(CborType::String("Signature".to_string()));
+    // TODO: add the protected body header here.
+    sig_structure.push(CborType::Bytes(vec![]));
+
+    // Encode values into the protected_sig_header, which is a byte array
+    // encoding a map.
+    if alg != CoseSignatureType::ES256 {
+        // TODO: we only accept ES256 for now.
+        return Vec::new();
+    }
+    let mut header_map: BTreeMap<CborType, CborType> = BTreeMap::new();
+    header_map.insert(CborType::Integer(1), CborType::SignedInteger(-7));
+    let header_map = CborType::Map(header_map).serialize();
+    sig_structure.push(CborType::Bytes(header_map));
+
+    // XXX: We don't hande external_aad here.
+    sig_structure.push(CborType::Bytes(vec![]));
+
+    sig_structure.push(CborType::Bytes(payload.to_vec()));
+
+    let sig_structure = CborType::Array(sig_structure);
+    sig_structure.serialize()
+}
+
+fn build_spki(key: &[u8]) -> Vec<u8> {
+    let mut spki = vec![0x30, 0x59, 0x30, 0x13, 0x06, 0x07, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x02,
+                        0x01, 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x03, 0x01, 0x07, 0x03,
+                        0x42, 0x00];
+    spki.extend_from_slice(key);
+    spki
+}
+
+/// The key has to be an uncrompressed curve point on the point specified in alg.
+pub fn sign(
```

Comment:
This seems like an API we're really only implementing for the purposes of testing the verification side of the API. Should this just go in some internal test module?